src/viam/sdk/common/proto_type.hpp

```diff
@@ -27,64 +27,82 @@ struct get_helper {
     }
 };
 
-template <>
-struct get_helper<AttributeMap> {
-    template <typename V>
-    static AttributeMap get(V& v) {
-        auto* const result = boost::get<AttributeMap>(&v);
-        return result ? *result : nullptr;
-    }
-
-    template <typename V>
-    static std::shared_ptr<const AttributeMap::element_type> get(const V& v) {
-        auto* const result = boost::get<AttributeMap>(&v);
-        return result ? *result : nullptr;
-    }
-};
-
-}  // namespace prototype_details
+}  // namespace value_type_details
 
-class ProtoType {
+class AttributeMap {
    public:
-    ProtoType() {
-        proto_type_ = boost::blank();
-    }
-
-    explicit ProtoType(bool b) : proto_type_(std::move(b)) {}
-    explicit ProtoType(std::string s) : proto_type_(std::move(s)) {}
-    explicit ProtoType(const char* c) : proto_type_(std::string(c)) {}
-    explicit ProtoType(int i) : proto_type_(std::move(i)) {}
-    explicit ProtoType(double d) : proto_type_(std::move(d)) {}
-    explicit ProtoType(AttributeMap m) : proto_type_(std::move(m)) {}
-    explicit ProtoType(std::vector<std::shared_ptr<ProtoType>> v) : proto_type_(std::move(v)) {}
-    explicit ProtoType(const google::protobuf::Value& value);
-
-    google::protobuf::Value proto_value();
-    friend bool operator==(const ProtoType& lhs, const ProtoType& rhs);
-
-    template <typename T>
-    auto get() {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
-
-    template <typename T>
-    auto get() const {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
+    class value_type;
+
+    using map_type = std::unordered_map<std::string, value_type>;
+
+    class value_type {
+       public:
+        template <typename T>
+        auto get() {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        template <typename T>
+        auto get() const {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        using base_value_types_ =
+            boost::variant<boost::blank, std::string, const char*, int, float, double, bool>;
+        using value_types = boost::make_recursive_variant<
+            base_value_types_,
```

Comment:
Hmm... I agree that seeing `boost::variant` and then `boost::make_recursive_variant` on it immediately after is a little strange. My thinking though was that we still wanted the root type to be a variant. So in essence the three cases for a value type are 
```
| boost::variant<...>,
| std::vector<boost::variant<...>>,
| std::unordered_map<std::string, boost::variant<...>>,
```

In playing around with this definition I was able to build a `value_type` with raw values, with lists, with maps, with nested lists/maps, etc. etc. I also got the sense that `make_recursive_variant_over` was largely equivalent to `make_recursive_variant`, per description [here](https://www.boost.org/doc/libs/1_84_0/doc/html/boost/make_recurs_1_3_44_5_5_1_3.html).* That said, I can definitely believe I'm missing something and that use of `mpl`/`make_recursive_variant_over` would produce superior results!

* `type has behavior equivalent in every respect to [make_recursive_variant](https://www.boost.org/doc/libs/1_84_0/doc/html/boost/make_recursive_variant.html)< Sequence[0], Sequence[1], ... >::type (where Sequence[i] denotes the i-th element of Sequence), except that no upper limit is imposed on the number of types.`

Reply:
Yes, make_recursive_variant_over is equivalent. What I'm puzzled about is why there are two levels of variant, rather than just:

```
boost::make_recursive_variant<boost::blank, ..., bool, std::unordered_map<std::string, boost::recursive_variant_>, std::vector<boost::recursive_variant_>>
```

Why the second level of `variant`? I wouldn't expect it to be needed.

Reply:
Ahh I see, that makes sense. Yeah I don't think the second level of variant is needed, will fix.